### PR TITLE
feat: types based on inheritance

### PR DESCRIPTION
**Pull request by egorio**
When function param has missing type:
- trying to confirm that function is part of a class
- check all extended or inherited classes and interfaces to find the function with the same name 
- check if the function has a parameter with the same ~name~ index and type

Types `any` and `object` are ignored by default.

I'm pretty sure there are edge cases I didn't cover...

### CODE REVIEW

File: packages/codefixes/src/fixes/addMissingTypesBasedOnInheritance/index.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+          continue;
+        }
+
+        const resolvedType = checker.getTypeOfSymbolAtLocation(
+          ancestorMethod,
+          heritageClauseTypeNode
+        );
+
+        // Expect this to return type, but `any` is returner
+        // checker.typeToString(checker.getTypeOfSymbolAtLocation(ancestorMethod.getDeclarations()[0].parameters[0].symbol.valueDeclaration))
```
Comment (egorio) [inline packages/codefixes/src/fixes/addMissingTypesBasedOnInheritance/index.ts:44]: Expect to get type here, but having `any`. 

The `ancestorMethod` is part of `Animal`

File: packages/codefixes/src/codefixes.ts [hunk 2]
```diff
@@ -14,4 +15,7 @@ export const codefixes = new CodeFixesProvider([
     4082: new AddMissingExportCodeFix(),
   }),
   new TypescriptCodeFixCollection(),
+  new BaseCodeFixCollection({
+    7006: new AddMissingTypesBasedOnInheritanceCodeFix(),
+  }),
```
Comment (chadhietala) [inline packages/codefixes/src/codefixes.ts:20]: Is there a reason to add this collection here? Why wouldn't we have it in the first collection?

File: packages/codefixes/src/fixes/addMissingTypesBasedOnInheritance/fixtures/types-based-on-inheritance.ts.input [hunk 3]
```diff
@@ -0,0 +1,63 @@
+class Animal {
```
Comment (chadhietala) [inline packages/codefixes/src/fixes/addMissingTypesBasedOnInheritance/fixtures/types-based-on-inheritance.ts.input:1]: Can you add cases for:


Multiple Implements
```
interface Thing {
  thing(num: number): void
}

class MultipleImplements implements Wild, Thing {
  wild(place) { return 'foo' };
  thing(num) { console.log(num); };
}
```

Generic Interfaces
```
interface GenericThing<T = number> {
  thing(num: T): void
}

type Thing = { foo: number };

interface ConstrainedThing<T extends Thing> {
  thing(num: T): void
}

class GenericImpl implements GenericThing {
  thing(num) { console.log(num); };
}

class GenericImplTwo implements GenericThing<number> {
  thing(num) { console.log(num); };
}

class GenericImplTwo implements ConstrainedThing<{ foo: number, bar: string }> {
  thing(num) { console.log(num); };
}
```

Generic Classes
```
class GenericThing<T = number> {
  thing(num: T): void
}

type Thing = { foo: number };

class ConstrainedThing<T extends Thing> {
  thing(num: T) {
  
  }
}

class GenericImpl extends GenericThing {
  thing(num) { console.log(num); };
}

class GenericImplTwo extends GenericThing<number> {
  thing(num) { console.log(num); };
}

class GenericImplTwo extends ConstrainedThing<{ foo: number, bar: string }> {
  thing(num) { console.log(num); };
}
```

**Comment by chadhietala** [comment]
💥

File: packages/codefixes/src/codefixes.ts (hunk 2 above)
Comment (egorio) [reply packages/codefixes/src/codefixes.ts:20]: Codefixes are applied in the order defined there.
I want to try this one after TypeScrupt ones were tried to apply.

File: packages/codefixes/src/fixes/addMissingTypesBasedOnInheritance/fixtures/types-based-on-inheritance.ts.input (hunk 3 above)
Comment (egorio) [reply packages/codefixes/src/fixes/addMissingTypesBasedOnInheritance/fixtures/types-based-on-inheritance.ts.input:3]: Test cases added and handled :)

**Comment by lynchbomb** [comment]
Awesome work!

**Review by lynchbomb — APPROVED** [review]

**Pull request merged**